### ISSUE TITLE
Questions about Batch Normalization

### ISSUE THREAD

**Issue by mervess** [larq/larq#698]
Hi all,

1. I was wondering how you approached the _Shift based Batch Normalization(BN)_ solution to the traditional _BN_ layer mentioned in the original [paper](https://arxiv.org/pdf/1602.02830.pdf) in _larq_.
2. Without the _BN_ layer, the accuracy is almost at 0 level in any (CNN like) _larq_ model. What is the reason behind it in your opinion?

Thanks in advance.

**Comment by lgeiger** [comment]
> I was wondering how you approached the Shift based Batch Normalization(BN) solution to the traditional BN layer mentioned in the original paper in larq.

https://github.com/larq/compute-engine will automatically handle correct fusion of batch normalisation into our [binary convolution operation](https://docs.larq.dev/compute-engine/api/operators/#lcebconv2d) when converting the larq model into an inference graph. In the case of two binary convolutions with a batch normalization in between, we will automatically transform a standard Keras batch normalization into a threshold computation if converted with [`experimental_enable_bitpacked_activations=True`](https://docs.larq.dev/compute-engine/api/python/#convert_keras_model). This means there will be no higher precision operations needed in the case of `lq.conv -> keras.batchnorm -> lq.conv`.

> Without the BN layer, the accuracy is almost at 0 level in any (CNN like) larq model. What is the reason behind it in your opinion?

The outputs of a normal binarized convolution are integer values:
```python
import tensorflow as tf
import larq as lq

x = tf.random.normal(shape=(100, 32, 32, 64))
y = lq.layers.QuantConv2D(64, 3, input_quantizer="ste_sign", kernel_quantizer="ste_sign")(x).numpy()

print(y.min(), y.mean(), y.max(), y.var())  # -118.0 -0.018400695 126.0 575.9885
```
These large numerical values are always problematic when training neural networks. But in particular in BNNs when followed by a default [STE Sign function](https://docs.larq.dev/larq/api/quantizers/#stesign) training will probably collapse. This ist due to the fact that the STE sign has zero gradients outside of `(-1, 1)`. As most activations without some sort of normalisation or rescaling will be outside this range the earlier layers might not even receive any useful gradient signal during training.

**Comment by mervess** [comment]
Many thanks for the detailed answer.
